corporate/lib/stripe.py

```diff
@@ -629,6 +632,7 @@ def process_initial_upgrade(
     realm = user.realm
     customer = update_or_create_stripe_customer(user)
     assert customer.stripe_customer_id is not None  # for mypy
+    assert customer.realm is not None
     ensure_realm_does_not_have_active_plan(customer.realm)
```

Comment:
This is great for this commit.

Just to comment a bit on the next stage of refactoring once we merge this: We're probably going to want to take methods like this and replace them with calls to methods like `ensure_no_active_plan(customer)`, with the goal of having this code not reference the realm object at all.

Meanwhile, the `ensure_no_active_plan` method could just have an if/else construction, or for now only support the `customer.realm!=None` case; the point is to start containing the code that needs to know about realms/servers into a set of functions that can execute the relevant steps.

Actually, one plausible theory for how to organize this is to have the methods that need to know about the realm/server difference be primarily class methods on Customer (which may just be if/else blocks); so this would call `customer.assert_no_active_plan()`, and later `customer.do_change_plan_type(...)`, `customer.get_latest_seat_count()`, etc.

But in any case I think that direction will lead to reasonably happy abstractions here.

Reply:
@timabbott I really like the idea of changing methods such as `ensure_no_active_plan` and `do_change_plan_type` to be class methods on `Customer`. I do have a couple of observations:
- `do_change_plan_type` is more so a method that updates `Realm.plan_type` and it doesn't really deal with the `Customer` or `CustomerPlan` models. Should we be adding a `plan_type` attribute to `ZulipRemoteServer` as well and then change pass in the `Customer` to `do_change_plan_type`?
- `ensure_no_active_plan` is also similar. We tend to get the `Customer` filtered by `Realm` and then the `CustomerPlan` filtered by `Customer` and then we make the required assertion. There are cases where we have access to `user.realm` and we pass that to the method. In those cases, should we get the `Customer` first and then pass it to `ensure_no_active_plan`? This seems a bit cumbersome to me since we already have access to the realm via `UserProfile`?

I'll do this refactoring in a follow-up commit, let me know what you think! Thanks!

Reply:
I think likely we'll end up adding a `plan_type` attribute to `ZulipRemoteServer`, yes.  Though I'm not sure what the best model would be; we might just want to rename the existing `do_change_plan_type` to `do_change_realm_plan_type`, and have the Customer method just call one of `do_change_realm_plan_type` and a new `do_change_remote_server_plan_type`.

I don't have a clear idea for the precise architecture one wants, beyond that it'd be nice to potentially contain logic around checking which of `realm` and `remote_server` is not None inside the `Customer` class.